src/virtual-machines/source-1-with-mark-region-gc.js

```diff
@@ -0,0 +1,1888 @@
+    HEAP[A + BLOCK_STATE_SLOT] = MARKED;
+    // mark node's start line to end line
+    A = SCAN;
+    GET_LINE();
+    A = RES;
+    while (
+      HEAP[A + LINE_ADDRESS_SLOT] + LINE_SIZE <=
+      SCAN + HEAP[SCAN + SIZE_SLOT]
+    ) {
+      HEAP[A + LINE_MARK_SLOT] = MARKED;
+      A = A + LINE_BK_SIZE;
+    }
+    for (
+      I = HEAP[SCAN + FIRST_CHILD_SLOT];
+      I <= HEAP[SCAN + LAST_CHILD_SLOT];
+      I = I + 1
+    ) {
+      A = HEAP[SCAN + I]; // address of child
+      GET_LINE();
+      A = HEAP[SCAN + I]; // address of child
+      if (
+        // child is marked and corresponding line also marked
+        HEAP[A + MARK_SLOT] === MARKED &&
+        HEAP[RES + LINE_MARK_SLOT] === MARKED
+      ) {
+      } else {
+        PUSH_RTS();
+      }
+    }
+  }
+}
+
+// set all blocks to be recyclable for now
+// expects hole-size in K
+function FREE_REGION() {
+  // implement free region
+  for (I = 0; I < NUMBER_OF_BLOCKS; I = I + 1) {
+    if (HEAP[I * BLOCK_SIZE + MARK_SLOT] === UNMARKED) {
+      HEAP[I * BLOCK_SIZE + BLOCK_STATE_SLOT] = FREE;
+    } else {
+      for (
+        // line pseudo node address in SCAN
+        SCAN = HEAP[I * BLOCK_SIZE + FIRST_CHILD_SLOT];
+        SCAN < HEAP[I * BLOCK_SIZE + LAST_CHILD_SLOT];
+        SCAN = SCAN + LINE_BK_SIZE
+      ) {
+        // unmark line
+        HEAP[SCAN + LINE_MARK_SLOT] = UNMARKED;
+        // free line
+        HEAP[SCAN + LINE_LIMIT_SLOT] = HEAP[SCAN + LINE_ADDRESS_SLOT];
+      }
+      // set block to recyclable
+      HEAP[I * BLOCK_SIZE + BLOCK_STATE_SLOT] = RECYCLABLE;
+    }
+    // unmark whole block
+    HEAP[I * BLOCK_SIZE + MARK_SLOT] = FREE;
+  }
+  // allocate new bump head and tail
+  for (I = 0; I < NUMBER_OF_BLOCKS; I = I + 1) {
+    if (
+      // if block is free or recyclable,
+      HEAP[I * BLOCK_SIZE + BLOCK_STATE_SLOT] === FREE ||
+      HEAP[I * BLOCK_SIZE + BLOCK_STATE_SLOT] === RECYCLABLE
+    ) {
+      // find k-sized hole
+      SCAN = HEAP[I * BLOCK_SIZE + FIRST_CHILD_SLOT];
+      while (SCAN < HEAP[I * BLOCK_SIZE + LAST_CHILD_SLOT]) {
+        if (HEAP[SCAN + LINE_LIMIT_SLOT] === HEAP[SCAN + LINE_ADDRESS_SLOT]) {
+          // if line is free/empty
+          // set bump head to end of previous line
+          BUMP_HEAD = HEAP[SCAN - LINE_BK_SIZE + LINE_ADDRESS_SLOT];
+          A = BUMP_HEAD + K;
+          GET_LINE();
+          if (HEAP[RES + LINE_LIMIT_SLOT] - BUMP_HEAD >= K) {
+            BUMP_TAIL = HEAP[RES + LINE_LIMIT_SLOT];
+            return undefined;
+          } else {
+            SCAN = SCAN + LINE_BK_SIZE;
+          }
+        } else {
+          SCAN = SCAN + LINE_BK_SIZE;
+        }
+      }
+    } else {
+    }
+  }
+}
+
+// expects object address in A
+// returns line node address in RES
+function GET_LINE() {
+  // block address in C
+  C = math_floor(A / BLOCK_SIZE) * BLOCK_SIZE;
+  // abuse RES to store start of addressable space
+  RES = HEAP[HEAP[C + FIRST_CHILD_SLOT] + LINE_ADDRESS_SLOT];
```

Comment:
```suggestion
  RES = HEAP[C + HEAP[C + FIRST_CHILD_SLOT] + LINE_ADDRESS_SLOT];
```
Is there a missing C?